### PR TITLE
Update logging secure forward

### PR DESCRIPTION
**Pull request by mburke5678**
Per slack convo with @richm 
https://coreos.slack.com/archives/CB3HXM2QK/p1571171589181800?thread_ts=1571167864.174500&cid=CB3HXM2QK

* Added a full `secure_forward` section from the `fluentd` configmap
* Added a step to update the `secure_forward` with the cert file and private key
* Added an example of the secret patch command with the same cert file and private key in new step.

### CODE REVIEW

**Comment by richm** [comment]
While you're in here, remove the Note about rsyslog.  In fact, these directions only apply for doing Fluentd -> Fluentd forwarding, since only Fluentd supports the `forward`/`secure_forward` protocol.

File: modules/efk-logging-fluentd-external.adoc [hunk 1]
```diff
@@ -51,10 +76,45 @@ $ oc edit configmap/fluentd -n openshift-logging
...
+    # </server>
+    # <server>
+      # ip address to connect
+    #   host 203.0.113.8
+      # specify hostlabel for FQDN verification if ipaddress is used for host
+    #   hostlabel server.fqdn.example.com
+    # </server>
+    # </store>
+----
+<1> Add the path to your certificate.
```
Comment (richm) [inline modules/efk-logging-fluentd-external.adoc]: your CA certificate

File: modules/efk-logging-fluentd-external.adoc [hunk 2]
```diff
@@ -51,10 +76,45 @@ $ oc edit configmap/fluentd -n openshift-logging
...
+    # @type secure_forward
+
+    # self_hostname ${hostname}
+    # shared_key <SECRET_STRING>
+
+    secure yes
+    # enable_strict_verification yes
+
+    ca_cert_path /etc/fluent/keys/ca.crt <1>
+    ca_private_key_path /etc/fluent/keys/privkey.pem <2>
```
Comment (richm) [inline modules/efk-logging-fluentd-external.adoc]: if you use mTLS (aka client cert auth with priv key) you will need to specify both the client cert and key using https://docs.fluentd.org/output/forward#tls_client_cert_path and https://docs.fluentd.org/output/forward#tls_client_private_key_path

Comment (richm) [reply modules/efk-logging-fluentd-external.adoc]: See also https://docs.fluentd.org/output/forward#how-to-connect-to-a-tls-ssl-enabled-server
that is, you can use tls without using tls_client_cert_path and tls_client_private_key_path - however, in order to use tls_client_cert_path and tls_client_private_key_path, you must also use tls

File: modules/efk-logging-fluentd-external.adoc [hunk 3]
```diff
@@ -51,10 +76,45 @@ $ oc edit configmap/fluentd -n openshift-logging
...
+    # <server>
+      # ip address to connect
+    #   host 203.0.113.8
+      # specify hostlabel for FQDN verification if ipaddress is used for host
+    #   hostlabel server.fqdn.example.com
+    # </server>
+    # </store>
+----
+<1> Add the path to your certificate.
+<2> Add the path to your private key.
```
Comment (richm) [inline modules/efk-logging-fluentd-external.adoc]: probably omit this and say something like "If you want to use mTLS (client cert auth), see the https://docs.fluentd.org/output/forward page for information about client cert and key parameters and other settings.".

**Comment by mburke5678** [comment]
@anpingli Please take a look. Thank you!!

**Comment by vikram-redhat** [comment]
@mburke5678 should this be in master?

**Comment by mburke5678** [comment]
@richm Do these changes apply to 4.2 and 4.3?

**Comment by richm** [comment]
It is confusing.  There are a lot of references to using the `@type secure_forward` plugin, and to using the secure_forward feature of fluentd 0.12, which is only applicable to 3.11.

What is the intention of this documentation for 4.x?  Do we need to have any references to "This is how you do it if you use secure_forward and/or fluentd 0.12" in our 4.x docs?

**Comment by mburke5678** [comment]
@richm I would think we want to document the _appropriate_ version of the plugin for 4.x. Do we not have that here?

File: modules/efk-logging-fluentd-external.adoc [hunk 4]
```diff
@@ -17,24 +17,44 @@ ifdef::openshift-enterprise[]
...
 The logging deployment provides a `secure-forward.conf` section in the Fluentd configmap
-for configuring the external aggregator:
+in the `openshift-logging` project for configuring the external aggregator:
+
+.Sample `secure-forward.conf` section
+[source,yaml]
+----
+ secure-forward.conf: |
+    # <store>
+    # @type secure_forward
```
Comment (richm) [inline modules/efk-logging-fluentd-external.adoc]: This sample config is for 3.11.  Use https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/configs.d/user/secure-forward.conf as the sample config for 4.x

File: modules/efk-logging-fluentd-external.adoc [hunk 5]
```diff
@@ -51,10 +71,46 @@ $ oc edit configmap/fluentd -n openshift-logging
 ----
 <1> Enter the name, host, and port for your external Fluentd server.
 
+. Add the path to your certificate and private key to the `secure-forward.conf` section:
++
+[source,yaml]
+----
+  secure-forward.conf: |
+    # <store>
+    # @type secure_forward
```
Comment (richm) [inline modules/efk-logging-fluentd-external.adoc]: This config is for 3.11.  Use https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/configs.d/user/secure-forward.conf as the sample config for 4.x

**Comment by richm** [comment]
Also, the final section starting with "The following is an example of a in_forward configuration for Fluentd 0.12:" - this is for _output_ not _input_. I think we should just remove this section entirely, and let the user use the external fluentd documentation for reference.  If we need to have an example config then use
```
<source>
  @type forward

  <security>
    self_hostname ${HOSTNAME}
    shared_key the-shared-key-used-with-client
  </security>
  bind 0.0.0.0
  port 24284
</source>
```
for fluentd 1.x and
```
<source>
  @type secure_forward

  self_hostname ${HOSTNAME}
  bind 0.0.0.0
  port 24284

  shared_key the-shared-key-used-with-client

  secure no
</source>
```
for fluentd 0.12

File: modules/efk-logging-fluentd-external.adoc (hunk 4 above)
Comment (mburke5678) [reply modules/efk-logging-fluentd-external.adoc]: The sample config in the docs is what is in the 4.2 Fluentd configmap. The user should replace the default `secure-forward.conf` in the configmap with the config in the link?

Comment (richm) [reply modules/efk-logging-fluentd-external.adoc]: > The sample config in the docs is what is in the 4.2 Fluentd configmap

Ah, ok.  That is wrong.  We need to fix that.   https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/configs.d/user/secure-forward.conf is what should be used there instead.

**Comment by mburke5678** [comment]
@anpingli Is this an issue you can review?

**Comment by anpingli** [comment]
I'd like use the external fluentd documentation for reference https://docs.fluentd.org/output/forward

**Comment by mburke5678** [comment]
I am proposing PR https://github.com/openshift/openshift-docs/pull/17558 in favor of the current PR, as I think I have a better understanding of how to use the forward plugins. PTAL

**Pull request closed without merging**